pwndbg/commands/__init__.py

```diff
@@ -88,82 +101,133 @@ class Command:
+        assert parser
+        self.parser = parser
+        # Sets self.help_str and self.description (among other stuff).
+        self.initialize_parser()
+
+        # Let the debugger and pwndbg global state know about it.
+        self.register_command()
+
+        # For commands like hexdump where you get new output from
+        # continuous invocations.
+        self.repeat = False
+
+    def register_command(self):
+        """
+        Register this object command with the underlying debugger
+        and update pwndbg global state to know about this command.
+        """
 
         def _handler(_debugger, arguments, is_interactive):
             self.invoke(arguments, is_interactive)
 
-        self.handle = pwndbg.dbg.add_command(command_name, _handler, doc)
-        self.function = function
-
-        if command_name in command_names:
-            raise Exception(f"Cannot add command {command_name}: already exists.")
-        if (
-            command_name in GDB_BUILTIN_COMMANDS
-            and command_name not in self.builtin_override_whitelist
-            and not pwndbg_is_reloading
-        ):
-            raise Exception(f'Cannot override non-whitelisted built-in command "{command_name}"')
+        # Tell the debugger about it.
+        self.handle = pwndbg.dbg.add_command(self.command_name, _handler, self.help_str)
 
-        command_names.add(command_name)
+        command_names.add(self.command_name)
         commands.append(self)
 
-        functools.update_wrapper(self, function)
-        self.__name__ = command_name
+    def initialize_parser(self):
+        # Set parser.prog so the help is generated properly.
+        self.parser.prog = self.command_name
 
-        self.repeat = False
+        # Generate command help
+        file = io.StringIO()
+        self.parser.print_help(file)
+        file.seek(0)
+        self.help_str = file.read()
 
-    def split_args(self, argument: str) -> Tuple[List[str], Dict[Any, Any]]:
-        """Split a command-line string from the user into arguments.
+        # Used by `pwndbg [filter]`
+        assert (
+            self.parser.description
+            and self.parser.description.strip()
+            and "A command must contain a description."
+        )
+        self.description = self.parser.description.strip()
 
-        This is only used by pwndbg/commands/shell.py which is deprecated.
-        Usually _ArgparsedCommand.split_args is called.
+        # We want to run all integer and otherwise-unspecified arguments
+        # through fix() so that GDB parses it.
+        # FIXME: this is weird
+        for action in self.parser._actions:
+            if isinstance(action, argparse._SubParsersAction):
+                action.type = str
+            if action.dest == "help":
+                continue
+            if action.type is int:
+                action.type = fix_int_reraise_arg
+            if action.type is None:
+                action.type = fix_reraise_arg
```

Comment:
Should the `action.dest == "help"` be the first if condition here?

Also shouldn't some of those be `elif`?